File: src/lib/data_types.ts
/**
 * This file contains TS type definitions of data sent between frontend and backend portions of the tauri app
 * These types should be changed alongside their Rust counterparts in `src-tauri/src/data_types.rs`
 */


/** Loose TS version of Rust struct `data_types::SysInfo` */
export interface SysInfo {
  os_info: {
    os: 'linux' | 'windows' | 'macos' | 'other';
    distro_name?: string;
    desktop_environment?: 'kde' | 'gnome' | 'hyprland' | { other: string } | 'unknown';
    session_type?: 'wayland' | 'x11' | 'tty' | { other: string };
  };
  support: 'supported' | 'unsupported' | 'broken' | 'unknown';
}

export type ActionType = 'sleep' | 'key' | 'craft';
export type PlayerEventType = 'update' | 'running' | 'paused' | 'stopped' | 'error';


// Macro and Action types
export type ModifierKey = 'shift' | 'ctrl' | 'alt' | 'super';
export interface KeyCombo {
  modifiers?: ModifierKey[];
  key: string;
}
interface BaseAction {
  actionId: number;
  action: ActionType;
}
export interface SleepAction extends BaseAction {
  action: 'sleep';
  duration_ms: number;
}
export interface KeyAction extends BaseAction {
  action: 'key';
  key: KeyCombo | undefined | null;
}
export interface CraftAction extends BaseAction {
  action: 'craft';
}
export type Action = SleepAction | KeyAction | CraftAction;
/** TS version of Rust struct `data_types::Macro` */
export interface RustMacro {
  id: number;   // TODO move ID assignment to Rust
  actions: Array<Omit<SleepAction, 'actionId'> | Omit<KeyAction, 'actionId'> | Omit<CraftAction, 'actionId'>>;
  loops: number;
}


// Commands
// export type PlayerCommandType = 'execute' | 'pause' | 'resume' | 'stop';
// interface BasePlayerCommand {
//   command: PlayerCommandType;
// }
// interface ExecuteCommand extends BasePlayerCommand {
//   command: 'execute';
//   actions: Action[];
// }
// interface PauseCommand extends BasePlayerCommand {
//   command: 'pause';
// }
// interface ResumeCommand extends BasePlayerCommand {
//   command: 'resume';
// }
// interface StopCommand extends BasePlayerCommand {
//   command: 'stop';
// }
// export type PlayerCommand = ExecuteCommand | PauseCommand | ResumeCommand | StopCommand;


// Event types
export interface ProgressInfo {
  action_index: number;     // integer in (0, len(actions))
  action_progress: number;  // float in (0.0, 1.0)
  loop_count: number;       // integer in (0, loops)
}
interface BaseMacroEvent {
  event_type: PlayerEventType;
  /** Macro id */
  id: number;
}
interface BaseUpdateEvent extends BaseMacroEvent {
  event_type: 'update' | 'running' | 'paused' | 'error';
  progress: ProgressInfo;
}
export interface UpdateEvent extends BaseUpdateEvent {
  event_type: 'update';
}
export interface RunningEvent extends BaseUpdateEvent {
  event_type: 'running';
}
export interface PausedEvent extends BaseUpdateEvent {
  event_type: 'paused';
}
export interface StoppedEvent extends BaseMacroEvent {
  event_type: 'stopped';
}
export interface ErrorEvent extends BaseUpdateEvent {
  event_type: 'error';
  error: string;
}
export type MacroEvent = UpdateEvent | RunningEvent | PausedEvent | StoppedEvent | ErrorEvent;

File: src-tauri/src/data_types.rs
/// This file contains Rust type definitions of data sent between frontend and backend portions of the tauri app
/// These types should be changed alongside their TS counterparts in `src/lib/data_types.ts`

// SysInfo types
#[derive(serde::Serialize, Clone)]
#[serde(rename_all = "lowercase")]
pub enum SessionType {
    Wayland,
    X11,
    Tty,
    Other(String),
}
#[derive(serde::Serialize, Clone)]
#[serde(rename_all = "lowercase")]
pub enum DEInfo {
    KDE,
    Gnome,
    Hyprland,
    Other(String),
    Unknown,
}
#[derive(serde::Serialize, Clone)]
#[serde(tag = "os", rename_all = "lowercase")]
pub enum OsInfo {
    Linux {
        distro_name: String,
        desktop_environment: DEInfo,
        session_type: SessionType,
    },
    Windows,
    MacOS,
    Other,
}
#[derive(serde::Serialize, Clone)]
#[serde(rename_all = "lowercase")]
pub enum Support {
    Supported,
    Unsupported,
    // Broken,
    Unknown,
}
#[derive(serde::Serialize, Clone)]
pub struct SysInfo {
    pub os_info: OsInfo,
    pub support: Support,
}

// Macro options
#[derive(serde::Deserialize, Copy, Clone, std::fmt::Debug)]
#[serde(rename_all = "lowercase")]
pub enum WindowsWIDMode {
    None,
    Title,
    Process,
}
#[derive(serde::Deserialize, Copy, Clone, std::fmt::Debug)]
#[serde(rename_all = "lowercase")]
pub enum WindowsWinMatchMode {
    Simple,
    Regex,
}
#[derive(serde::Deserialize, std::fmt::Debug)]
#[serde(tag = "type", rename_all = "lowercase")]
pub enum MacroOptions {
    Hyprland { window_identifier: String },
    Windows {
        /// Determines if the input should go to a specific window and how that window will be identified
        window_id_mode: WindowsWIDMode,
        /// Window identifier string
        window_id_str: String,
        /// Window identifier match mode
        match_mode: WindowsWinMatchMode,
        /// Whether the macro executor should focus the target window or wait for it to get focused
        auto_focus: bool,
    }
}

// Macro action types
#[derive(serde::Deserialize, std::fmt::Debug)]
#[serde(rename_all = "lowercase")]
pub enum ModifierKey {
    Shift,
    Ctrl,
    Alt,
    Super,
}
#[derive(serde::Deserialize, std::fmt::Debug)]
pub struct KeyCombo {
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub modifiers: Vec<ModifierKey>,
    pub key: String,
}
#[derive(serde::Deserialize, std::fmt::Debug)]
#[serde(tag = "action", rename_all = "lowercase")]
pub enum MacroAction {
    Sleep { duration_ms: u64 },
    Key { key: Option<KeyCombo> },
    Craft,
}

// Macro
#[derive(serde::Deserialize)]
pub struct Macro {
    pub id: u32,
    pub actions: Vec<MacroAction>,
    pub loops: u32,
    pub options: MacroOptions,
}

// Event types
#[derive(serde::Serialize, Clone)]
pub struct MacroProgress {
    /// index of the currently running action
    pub action_index: u32,
    /// current action's progress as a floating point value between 0.0 and 1.0
    pub action_progress: f64,
    /// current loop number
    pub loop_count: u32,
}
#[derive(serde::Serialize, Clone)]
#[serde(tag = "event_type", rename_all = "lowercase")]
pub enum MacroEvent {
    Update { id: u32, progress: MacroProgress },
    Running { id: u32, progress: MacroProgress },
    Paused { id: u32, progress: MacroProgress },
    Stopped { id: u32 },
    Error { id: u32, progress: MacroProgress, error: String },
}
